### PR TITLE
Correct and expand documentation for the ImageDecoder API.

### PR DESCRIPTION
**Pull request by dalecurtis**
#### Summary
Corrects and expands the documentation for the ImageDecoder API.

#### Motivation
Helps developers and users understand the API better.

#### Supporting details
https://w3c.github.io/webcodecs/
https://w3c.github.io/webcodecs/samples/

I'm an author of the ImageDecoder API :)

#### Related issues
n/a

#### Metadata

- [x] Adds a new document
- [x] Rewrites (or significantly expands) a document
- [x] Fixes a typo, bug, or other error

### CODE REVIEW

File: files/en-us/web/api/imagedecoder/close/index.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ browser-compat: api.ImageDecoder.close
 ---
 {{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
 
-The **`close()`** method of the {{domxref("ImageDecoder")}} interface ends all pending work and releases system resources.
+The **`close()`** method of the {{domxref("ImageDecoder")}} interface ends all pending work and releases system resources. Subsequent calls to `decode()` will throw an `InvalidStateError` exception.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/close/index.md:15]: Exceptions should not be described in the intro, but in a specific section:
```suggestion
The **`close()`** method of the {{domxref("ImageDecoder")}} interface ends all pending work and releases system resources.
```

and the exception section should be between _Return value_ and _Examples_ and be:
```

### Exceptions

- `InvalidStateError` {{domxref("DOMException")}}
  - : Thrown if the decoder is already closed.
 
```

File: files/en-us/web/api/imagedecoder/imagedecoder/index.md [hunk 2]
```diff
@@ -24,22 +24,22 @@ new ImageDecoder(init)
 - `init`
   - : An object containing the following members:
     - `type`
-      - : A string containing the MIME type of the image file to be decoded.
+      - : A string containing the [MIME type](https://en.wikipedia.org/wiki/Media_type) of the image file to be decoded.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/imagedecoder/index.md:27]: We have an internal link.

```suggestion
      - : A string containing the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) of the image file to be decoded.
```

File: files/en-us/web/api/imagedecoder/index.md [hunk 3]
```diff
@@ -21,20 +21,85 @@ The **`ImageDecoder`** interface of the {{domxref('WebCodecs API','','','true')}
 ## Properties
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/index.md:21]: On line 12, please change the incorrect `{{DefaultAPISidebar("WebCodecs API")}}` to `{{APIRef("WebCodecs API")}}`. Thank you.

File: files/en-us/web/api/imagedecoder/index.md [hunk 4]
```diff
@@ -21,20 +21,85 @@ The **`ImageDecoder`** interface of the {{domxref('WebCodecs API','','','true')}
...
+  - : Returns a {{jsxref("DOMString")}} reflecting the [mime type](https://en.wikipedia.org/wiki/Media_type) configured during construction.
 
 ## Methods
 
 - {{domxref("ImageDecoder.close()")}}
   - : Ends all pending work and releases system resources.
 - {{domxref("ImageDecoder.decode()")}}
   - : Enqueues a control message to decode the frame of an image.
+- {{domxref("ImageDecoder.isTypeSupported()")}}
+  - : Indicates if the provided [mime type](https://en.wikipedia.org/wiki/Media_type) is supported for unpacking and decoding.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/index.md:39]: Create a new separate section (`## Static methods`) for static methods.

File: files/en-us/web/api/imagedecoder/index.md [hunk 5]
```diff
@@ -21,20 +21,85 @@ The **`ImageDecoder`** interface of the {{domxref('WebCodecs API','','','true')}
...
 - {{domxref("ImageDecoder.reset()")}}
-  - : Resets all states including configuration, control messages in the control message queue, and all pending callbacks.
+  - : Aborts all pending `decode()` operations.
+
+## Examples
+
+Given an {{HTMLElement("canvas")}} element:
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/index.md:45]:
```suggestion
Given a {{HTMLElement("canvas")}} element:
```

File: files/en-us/web/api/imagedecoder/istypesupported/index.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+---
+title: ImageDecoder.isTypeSupported
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/istypesupported/index.md:2]:
```suggestion
title: ImageDecoder.isTypeSupported()
```

File: files/en-us/web/api/imagedecoder/istypesupported/index.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+page-type: web-api-static-method
+tags:
+  - API
+  - Method
+  - Reference
+  - istypesupported
+  - ImageDecoder
+browser-compat: api.ImageDecoder.isTypeSupported
+---
+{{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/istypesupported/index.md:13]:
```suggestion
{{securecontext_header}}{{APIRef("WebCodecs API")}}
```

File: files/en-us/web/api/imagedecoder/index.md [hunk 8]
```diff
@@ -21,20 +21,85 @@ The **`ImageDecoder`** interface of the {{domxref('WebCodecs API','','','true')}
...
 
 - {{domxref("ImageDecoder.complete")}}{{ReadOnlyInline}}
-  - : Returns a {{jsxref("boolean")}} indicating whether the data is completely buffered.
+  - : Returns a {{jsxref("boolean")}} indicating whether encoded data is completely buffered.
 - {{domxref("ImageDecoder.completed")}}{{ReadOnlyInline}}
   - : Returns a {{jsxref("Promise")}} that resolves once `complete` is true.
 - {{domxref("ImageDecoder.tracks")}}{{ReadOnlyInline}}
   - : Returns an {{domxref("ImageTrackList")}} object listing the available tracks and providing a method for selecting a track to decode.
+- {{domxref("ImageDecoder.type")}}{{ReadOnlyInline}}
+  - : Returns a {{jsxref("DOMString")}} reflecting the [mime type](https://en.wikipedia.org/wiki/Media_type) configured during construction.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/index.md:30]:
```suggestion
  - : Returns a {{jsxref("DOMString")}} reflecting the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) configured during construction.
```

File: files/en-us/web/api/imagedecoder/index.md (hunk 4 above)
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/index.md:39]:
```suggestion
  - : Indicates if the provided [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) is supported for unpacking and decoding.
```

File: files/en-us/web/api/imagedecoder/imagedecoder/index.md [hunk 9]
```diff
@@ -24,22 +24,22 @@ new ImageDecoder(init)
 - `init`
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/imagedecoder/index.md:24]: On line 12, please change the incorrect {{DefaultAPISidebar("WebCodecs API")}} to {{APIRef("WebCodecs API")}}. Thank you.

File: files/en-us/web/api/imagedecoder/decode/index.md [hunk 10]
```diff
@@ -28,7 +28,7 @@ decode(options)
     - `frameIndex`{{Optional_Inline}}
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/decode/index.md:28]: On line 13, please change the incorrect {{DefaultAPISidebar("WebCodecs API")}} to {{APIRef("WebCodecs API")}}. Thank you.

File: files/en-us/web/api/imagedecoder/completed/index.md [hunk 11]
```diff
@@ -12,7 +12,7 @@ browser-compat: api.ImageDecoder.completed
 ---
 {{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/completed/index.md:13]:
```suggestion
{{securecontext_header}}{{APIRef("WebCodecs API")}}
```

File: files/en-us/web/api/imagedecoder/complete/index.md [hunk 12]
```diff
@@ -12,7 +12,7 @@ browser-compat: api.ImageDecoder.complete
 ---
 {{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/complete/index.md:13]:
```suggestion
{{securecontext_header}}{{APIRef("WebCodecs API")}}
```

File: files/en-us/web/api/imagedecoder/close/index.md [hunk 13]
```diff
@@ -12,7 +12,7 @@ browser-compat: api.ImageDecoder.close
 ---
 {{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/close/index.md:13]:
```suggestion
{{securecontext_header}}{{APIRef("WebCodecs API")}}
```

File: files/en-us/web/api/imagedecoder/istypesupported/index.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+The **`ImageDecoder.isTypeSupported()`** static method checks if a given [mime type](https://en.wikipedia.org/wiki/Media_type) can be decoded by the user agent.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/istypesupported/index.md:15]:
```suggestion
The **`ImageDecoder.isTypeSupported()`** static method checks if a given [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) can be decoded by the user agent.
```

File: files/en-us/web/api/imagedecoder/istypesupported/index.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+### Parameters
+
+- `type`
+  - : The [mime type](https://en.wikipedia.org/wiki/Media_type) to check for decoding support.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/istypesupported/index.md:20]:
```suggestion
  - : The [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) to check for decoding support.
```

File: files/en-us/web/api/imagedecoder/reset/index.md [hunk 16]
```diff
@@ -12,7 +12,7 @@ browser-compat: api.ImageDecoder.reset
 ---
 {{securecontext_header}}{{DefaultAPISidebar("WebCodecs API")}}
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/reset/index.md:13]:
```suggestion
{{securecontext_header}}{{APIRef("WebCodecs API")}}
```

**Review by teoli2003 — CHANGES_REQUESTED** [review]
Thank you for the improvement, this is much better.

I've done technical suggestions to match our guidelines (and fix a few extra errors).

Unfortunately, there are a few places where the GitHub UI prevented me from directly adding the suggestion to the right place.

File: files/en-us/web/api/imagedecoder/imagedecoder/index.md (hunk 2 above)
Comment (dalecurtis) [reply files/en-us/web/api/imagedecoder/imagedecoder/index.md:27]: Done in all cases.

File: files/en-us/web/api/imagedecoder/index.md (hunk 3 above)
Comment (dalecurtis) [reply files/en-us/web/api/imagedecoder/index.md:21]: Resolved here and everywhere else.

**Comment by dalecurtis** [comment]
Thanks for the review. I think I addressed everything. I'll make a pass over the ImageTrack and ImageTrackList in a follow-up change.

File: files/en-us/web/api/imagedecoder/imagedecoder/index.md [hunk 17]
```diff
@@ -24,22 +24,22 @@ new ImageDecoder(init)
...
   - : An object containing the following members:
     - `type`
-      - : A string containing the MIME type of the image file to be decoded.
+      - : A {{domxref("USVString")}} containing the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) of the image file to be decoded.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/imagedecoder/index.md:27]: Sorry, I missed this one. We are retiring `DOMString`, `USVString` and `CSSOMString` as these names are invisible for a web dev, in the console for example.

(By retiring, I mean the pages will be deleted by the end of the week).
```suggestion
      - : A string containing the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) of the image file to be decoded.
```

File: files/en-us/web/api/imagedecoder/index.md [hunk 18]
```diff
@@ -21,11 +21,13 @@ The **`ImageDecoder`** interface of the {{domxref('WebCodecs API','','','true')}
...
 
 - {{domxref("ImageDecoder.complete")}}{{ReadOnlyInline}}
-  - : Returns a {{jsxref("boolean")}} indicating whether the data is completely buffered.
+  - : Returns a {{jsxref("boolean")}} indicating whether encoded data is completely buffered.
 - {{domxref("ImageDecoder.completed")}}{{ReadOnlyInline}}
   - : Returns a {{jsxref("Promise")}} that resolves once `complete` is true.
 - {{domxref("ImageDecoder.tracks")}}{{ReadOnlyInline}}
   - : Returns an {{domxref("ImageTrackList")}} object listing the available tracks and providing a method for selecting a track to decode.
+- {{domxref("ImageDecoder.type")}}{{ReadOnlyInline}}
+  - : Returns a {{domxref("DOMString")}} reflecting the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) configured during construction.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/index.md:30]:
```suggestion
  - : Returns a string reflecting the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) configured during construction.
```

File: files/en-us/web/api/imagedecoder/index.md [hunk 19]
```diff
@@ -21,11 +21,13 @@ The **`ImageDecoder`** interface of the {{domxref('WebCodecs API','','','true')}
 ## Properties
 
 - {{domxref("ImageDecoder.complete")}}{{ReadOnlyInline}}
-  - : Returns a {{jsxref("boolean")}} indicating whether the data is completely buffered.
+  - : Returns a {{jsxref("boolean")}} indicating whether encoded data is completely buffered.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/index.md:24]:
```suggestion
  - : Returns a boolean value indicating whether encoded data is completely buffered.
```

It returns a boolean value `true` or `false` and not a `Boolean` wrapper object.

File: files/en-us/web/api/imagedecoder/istypesupported/index.md [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+browser-compat: api.ImageDecoder.isTypeSupported
+---
+{{securecontext_header}}{{APIRef("WebCodecs API")}}
+
+The **`ImageDecoder.isTypeSupported()`** static method checks if a given [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) can be decoded by the user agent.
+
+### Parameters
+
+- `type`
+  - : A {{domxref("DOMString")}} containing the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) to check for decoding support.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/istypesupported/index.md:20]:
```suggestion
  - : A string containing the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) to check for decoding support.
```

File: files/en-us/web/api/imagedecoder/istypesupported/index.md [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+
+### Return value
+
+A {{jsxref("promise")}} that resolves with a {{jsxref("boolean")}} indicating whether images with a format of `type` can be decoded by the API.
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/istypesupported/index.md:24]:
```suggestion
A {{jsxref("promise")}} that resolves with a boolean value indicating whether images with a format of `type` can be decoded by the API.
```

File: files/en-us/web/api/imagedecoder/type/index.md [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+  - ImageDecoder
+browser-compat: api.ImageDecoder.type
+---
+{{securecontext_header}}{{APIRef("WebCodecs API")}}
+
+The **`type`** read-only property of the {{domxref("ImageDecoder")}} interface reflects the [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) configured during construction.
+
+## Value
+
+A {{domxref("DOMString")}} containing the configured [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types).
```
Comment (teoli2003) [inline files/en-us/web/api/imagedecoder/type/index.md:19]:
```suggestion
A string containing the configured [MIME type](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types).
```

**Review by teoli2003 — CHANGES_REQUESTED** [review]
Sorry, I missed the WebIDL types in my first pass. (I try to avoid to do multiple passes, but this was an oversight)

A few small changes and we can merge this!

**Comment by dalecurtis** [comment]
Thanks, committed all your suggestions! Thanks for making them inline.

**Review by teoli2003 — APPROVED** [review]
👍

**Comment by teoli2003** [comment]
Thanks a lot!

Looks good to me.

Congratulations upon your first merged commit here! Welcome aboard! :tada:

**Pull request merged**